[Tooling] Refactor code freeze lane and Fastfile cleanup

AliSoftware opened the pull request:
📘 Related to Project Thread: paaHJt-78B-p2

# What

This PR is only part one of a cleanup of the `Fastfile` in preparation for the release-on-CI migration project.

The main changes made by this PR are the refactoring of the `code_freeze` lane and by improving the way and time at which the backmerge PRs were created:

 - The `code_freeze` lane was initially split into sublanes `start_code_freeze` + `comment_on_pr` + `complete_code_freeze` that the `code_freeze` lane called in sequence. There was no real need for such a split, so now all the steps necessary for `code_freeze` are implemented directly in the `lane :code_frreze do…end`
 - The steps that were done during those lanes, especially the ones around commenting on still-opened PR, freezing the milestone, etc… were not made in the best order and not mirroring our approach in other repos
 - Some of those steps used a custom ruby implementation directly in the Fastfile while we have dedicated `release-toolkit` action for those. This was especially the case for:
    - commenting on still-opened PRs (raw API call to the `github_api` to find open PRs, and only commenting on them without updating their milestone)—now replaced by our `update_assigned_milestone` action
    - opening the backmerge PR (creation of intermediate branch during `code_freeze`, but creation of the backmerge PR using that intermediate branch… only at the end of the `build_and_upload_app_store_connect` run by CI, i.e. only after the build finished and was uploaded via `testflight`)—now replaced by our `create_release_backmerge_pull_request` action

# Next steps

I'll stack more PRs on top of this, using an iterative approach for the changes:
 - PR to reorder the lanes in the `Fastfile`: #2216
 - PR to migrate more lanes to use keyword syntax: #2217
 - PR to add the `release-pipelines/*.yml` pipelines that will allow triggering those lanes from CI via ReleaseV2
 - At some point, automate the submission for TestFlight builds (automate adding tester groups to TestFlight betas, automate submission to Apple for release builds)

I also plan to follow-up with updates to the ReleaseV2 scenario template too:

1. First diff: add manual tasks about the "next steps" to do after the Slack notification is sent.
    - In particular, the scenario currently has a task saying: "Watch for the #pocket-casts-dev-ios notification stating that the build has been submitted and follow the mentioned steps.", and the `slack_message` method indeed provide instructions on steps like "Please distribute it to testers and merge the PR"
    - I'll thus instead move those instructions to dedicated proper (manual) tasks in the ReleaseV2 scenario
2. One `release-pipelines/*.yml` are ready: replace the manual tasks which are asking to call the lanes on the RM's Terminal to automated tasks triggering those pipelines instead
3. Once submission of TestFlight builds is automated: diff to remove the manual task about adding tester groups / submitting build to Apple manually

AliSoftware committed:
Modernize `code_freeze`-related lanes

AliSoftware committed:
Move creation of backmerge PR to the triggering lanes

Instead of creating the intermediate branches in the triggering lanes line `code_freeze`/`new_beta_release`/`finalize_release`/`finalize_hotfix_release` but only creating the backmerge PR during the `build_and_upload_app_store_connect` lane run on CI and only after the build was finished and uploaded to TestFlight.

That way the backmerge PR is created as soon as the build is triggered, instead of having to wait for the CI build to finish to be able to review and merge it. This also means that the CI checks on the backmerge PR itself will start sooner rather than only starting after the beta/final release build is done, allowing parallelization of those 2 CI builds.

AliSoftware committed:
Use keyword parameters for `code_freeze` lane

dangermattic commented on fastlane/Fastfile:
```diff
@@ -302,16 +253,22 @@ platform :ios do
+  # - Open backmerge PR
+  # - Enables the GitHub branch protection for the new branch
+  # - Move all still-opened PRs targeting current milestone to the next one and Freezes the GitHub milestone
+  #  
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/689cde6eef99b303ccc74cf419b2f1b53392b113/fastlane/Fastfile#L266"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -1114,25 +1036,80 @@ inal copy, and try again.")
+  # - Loop though all PRs which were still open and were assigned the milestones of the code-frozen version,
+  #   and move them to the next milestone + leaving a PR comment about it
+  # - Add the ❄️ frozen marker to the milestone being frozen
+  def update_milestone()
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/689cde6eef99b303ccc74cf419b2f1b53392b113/fastlane/Fastfile#L1042"></span>Style/DefWithParentheses: Omit the parentheses in defs when the method doesn't accept any arguments.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -1114,25 +1036,80 @@ inal copy, and try again.")
+      - If this is not the first time you are running the release task (e.g. retrying because it failed on first attempt), the milestone might have already been closed and this error is expected.
+      - Otherwise if this is the first you are running the release task for this version, please investigate the error.
+    MESSAGE
+  
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/689cde6eef99b303ccc74cf419b2f1b53392b113/fastlane/Fastfile#L1088"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -1114,25 +1036,80 @@ inal copy, and try again.")
+    MESSAGE
+  
+    UI.error(error_message)
+  
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/689cde6eef99b303ccc74cf419b2f1b53392b113/fastlane/Fastfile#L1090"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -1114,25 +1036,80 @@ inal copy, and try again.")
+  
+    buildkite_annotate(style: 'warning', context: 'error-with-milestone', message: error_message) if is_ci
+  end
+  
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/689cde6eef99b303ccc74cf419b2f1b53392b113/fastlane/Fastfile#L1093"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/60">7.74</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware committed:
Fix rubocop violations

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -230,55 +230,6 @@ platform :ios do
   # Release Lanes
   #####################################################################################
 
-  # Executes the code freeze
-  #
-  # - Call start_code_freeze
-  # - Call complete_code_freeze
-  # - Go over PRs with the current milestone and add a comment to move them to the next
-  # - Open a PR targeting trunk
-  # - Send a message to Slack channel confirming the code freeze was finished
-  #
-  # @option [Boolean] skip_confirm (default: false) If true, avoids any interactive prompt
-  #
-  desc 'Executes all the code freeze steps'
-  lane :code_freeze do |options|
-    start_code_freeze(options)
-    comment_on_prs(milestone: release_version_current)
-    complete_code_freeze(options)
-  end
```
Instead of splitting the implementation of this `code_freeze` lane in separate `start_code_freeze` and `complete-code_freeze` lanes, I moved the whole implementation of the steps in the `code_freeze` lane directly, see [line 270 below](https://github.com/Automattic/pocket-casts-ios/pull/2215/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688R270)

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -230,55 +230,6 @@ platform :ios do
-    complete_code_freeze(options)
-  end
-
-  # Go through PRs and check their milestones, leaving a message if the target milestone is being frozen
-  def comment_on_prs(milestone:)
-    require_env_vars!('GITHUB_TOKEN')
-
-    version = milestone
-    UI.message("Checking open PRs pointing to #{version} milestone...")
-
-    github_api(
-      server_url: 'https://api.github.com',
-      api_token: get_required_env!('GITHUB_TOKEN'),
-      http_method: 'GET',
-      path: "/repos/#{GHHELPER_REPO}/pulls",
-      error_handlers: {
-        '*' => proc do |_result|
-          UI.important('⚠️ Failed checking open PRs, skipping this step.')
-        end
-      }
-    ) do |result|
-      result[:json].each do |pr|
-        next if pr['milestone'].nil?
-        next unless pr['milestone']['title'].include?(version)
-
-        comment_on_pr(
-          project: GHHELPER_REPO,
-          pr_number: pr['number'],
-          body: "\nThis PR is targeting #{version} which is entering code freeze. If this needs to be included in the #{version} release please merge it into `release/#{version}` if not please move it to the next milestone.",
-          reuse_identifier: 'code-freeze-in-progress'
-        )
-      end
-    end
-  end
```
This custom implementation was replaced by a call to our `update_assigned_milestone` action from `release-toolkit` instead.

The call to `update_assigned_milestone` is now wrapped inside [the `def update_milestone` helper function below](https://github.com/Automattic/pocket-casts-ios/pull/2215/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688R1042-R1067), which takes care of also adding the frozen marker ❄️  to the milestone after having moved the PRs, and handling potential exceptions (milestone already updated)

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -353,51 +309,27 @@ platform :ios do
-    after_confirming_push(push_merge_branch: true) do
+    after_confirming_push do
       trigger_beta_build(branch_to_build: release_branch_name)
+      create_backmerge_pr
```
Before this change, the `create_merge_release_branch` call only created the intermediate `merge/…` branch—and pushed it to remote during `after_confirming_push(push_merge_branch: true)`—but didn't open the backmerge PR at the time of the call to this `code_freeze` lane yet.

Instead, after the intermediate branch was created and pushed to the remote, the beta build was triggered, which ultimately called the `build_and_upload_app_store_connect` lane on CI, and that's [only in that lane that the backmerge PR was created](https://github.com/Automattic/pocket-casts-ios/pull/2215/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688L530-L541)—only after the CI had finished building the `.ipa` and uploaded it to TestFlight!

This PR changes the approach to not have the `build_and_upload_app_store_connect` lane be the one creating the backmerge PR, but instead have the lanes that are triggering those beta/release builds (namely `code_freeze` and `new_beta_release` which are calling `trigger_beta_build`, and `finalize_release` and `finalize_hotfix_release` which are calling `trigger_release_build`) be the ones also creating the PR at the same time.
This not only will allow the Release Manager to review the backmerge PR (and solve any potential conflict on it) sooner—without having to wait multiple minutes for the beta/release build to have finished building on CI for the PR to be created—but will also mean that the CI checks on that backmerge PR (unit tests, etc) will start sooner and run in parallel to the beta/release CI build, thus having that backmerge PR go green and ready to be merged sooner too.

iangmaia replied:
~For a future PR I guess: do we even need `after_confirming_push`?~ 
Nevermind. At first I didn't get why it wouldn't make sense to just push it, but it's a good idea to keep consistency with the confirmation checks we have.

AliSoftware replied:
Nope, I'm planning to remove this as well, but in a future PR indeed, as I didn't want to change too much at a time (it already gave me such headaches to wrap my head around those various helpers and custom implementations used in PC's Fastfile).

I'm probably gonna remove this when I'll get closer to finally have those lanes be run by CI, at which point all the confirmation points will be removed (or at least use `skip_confirm: true`). Even if we keep the possibility for confirmation (in case we want to still sometimes call those lanes from our Macs for debugging purposes) I'm planning to get rid of that `after_confirming_push` method and clean that up at some point anyway indeed to make the confirmation points, if we keep them, more consistent across the `Fastfile` and with our other repos.

AliSoftware committed:
Fix use of wrong constant name during copy/paste

danielebogo reviewed (approved):
LGTM!
Thanks for doing this!
Is there a way we can test it?

AliSoftware commented:
> LGTM!
> 
> Thanks for doing this!
> 
> Is there a way we can test it?

Sadly not really, not without doing a real code freeze and all.
So my hope is to land this today so that we can test this live during the code freeze from this Monday—as well as betas and final release during rest of the sprint.

I'll then obviously make myself available during that next sprint to jump on any fix that might be needed if anything goes wrong and the PCiOS release manager encounters issues due to my changes in that PR.

danielebogo commented:
@AliSoftware no problem! @SergioEstevao is the release manager for the next cycle but I'm happy to help with anything

iangmaia commented on fastlane/Fastfile:
```diff
@@ -613,10 +534,9 @@ platform :ios do
     commit_version_bump
```
Won't this `commit_version_bump` need to be pushed?

AliSoftware replied:
Ohh nice catch!!

_It was previous pushed by the call to `push_release_branches` but now that I removed that call from line 617 when I replaced the "only create intermediate branch in new_beta_release but create the backmerge PR only at the end of the build" with `create_backmerge_pr` and removed the calls that created and pushed the intermediate branches separately, I lost that part_

AliSoftware replied:
@iangmaia should be fixed in 34b70860d

_(I know, I used `after_confirming_push` while I plan to remove it 😅 … but we'll revisit that in a separate PR)_

AliSoftware committed:
Fix missing push of the version bump during `new_beta_release` before triggering the build

h/t @iangmaia for noticing this in https://github.com/Automattic/pocket-casts-ios/pull/2215\#discussion_r1778966319

iangmaia reviewed (approved):
:shipit:
